pkg/reconciler/kubequota/delegating_event_handler.go

```diff
@@ -0,0 +1,123 @@
+	"k8s.io/client-go/tools/cache"
+)
+
+// delegatingEventHandler multiplexes event handlers for multiple resource types and logical clusters.
+type delegatingEventHandler struct {
+	lock          sync.RWMutex
+	eventHandlers map[schema.GroupResource]map[logicalcluster.Name]cache.ResourceEventHandler
+}
+
+// newDelegatingEventHandler returns a new delegatingEventHandler.
+func newDelegatingEventHandler() *delegatingEventHandler {
+	return &delegatingEventHandler{
+		eventHandlers: map[schema.GroupResource]map[logicalcluster.Name]cache.ResourceEventHandler{},
+	}
+}
+
+// registerEventHandler registers an event handler, h, to receive events for the given resource/informer scoped only to
+// clusterName.
+func (d *delegatingEventHandler) registerEventHandler(
+	resource schema.GroupResource,
+	informer cache.SharedIndexInformer,
+	clusterName logicalcluster.Name,
+	h cache.ResourceEventHandler,
+) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	groupResourceHandlers, ok := d.eventHandlers[resource]
+	if !ok {
+		groupResourceHandlers = map[logicalcluster.Name]cache.ResourceEventHandler{}
+		d.eventHandlers[resource] = groupResourceHandlers
+
+		informer.AddEventHandler(d.resourceEventHandlerFuncs(resource))
+	}
+
+	groupResourceHandlers[clusterName] = h
+}
+
+// registerEventHandlerWithResyncPeriod registers an event handler, h, to receive events for the given resource/informer
+// scoped only to clusterName, with the given resync period.
+func (d *delegatingEventHandler) registerEventHandlerWithResyncPeriod(
+	resource schema.GroupResource,
+	informer cache.SharedIndexInformer,
+	clusterName logicalcluster.Name,
+	h cache.ResourceEventHandler,
+	resyncPeriod time.Duration,
+) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	groupResourceHandlers, ok := d.eventHandlers[resource]
+	if !ok {
+		groupResourceHandlers = map[logicalcluster.Name]cache.ResourceEventHandler{}
+		d.eventHandlers[resource] = groupResourceHandlers
+
+		informer.AddEventHandlerWithResyncPeriod(d.resourceEventHandlerFuncs(resource), resyncPeriod)
+	}
+
+	groupResourceHandlers[clusterName] = h
+}
+
+func (d *delegatingEventHandler) resourceEventHandlerFuncs(resource schema.GroupResource) cache.ResourceEventHandlerFuncs {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if h := d.getEventHandler(resource, obj); h != nil {
+				h.OnAdd(obj)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			if h := d.getEventHandler(resource, oldObj); h != nil {
+				h.OnUpdate(oldObj, newObj)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			if h := d.getEventHandler(resource, obj); h != nil {
+				h.OnDelete(obj)
+			}
+		},
+	}
+}
+
+// getEventHandler returns a cache.ResourceEventHandler for resource and the logicalcluster.Name for obj.
+func (d *delegatingEventHandler) getEventHandler(resource schema.GroupResource, obj interface{}) cache.ResourceEventHandler {
+	clusterName := clusterNameForObj(obj)
+	if clusterName.Empty() {
```

Comment:
am curious when this could happen. Wouldn't the same logic below keep working without this special casing?

Reply:
Ideally never, but probably was an overly aggressive optimization to avoid needing to grab the lock. Will simplify if you'd prefer?